Masonry isn't performing intrinsic sizing correctly.

stwrt opened the pull request:
#### 58b046421e15c1f20ea7cc5dc0e64bd58e11cd92
<pre>
Masonry isn&apos;t performing intrinsic sizing correctly.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270663">https://bugs.webkit.org/show_bug.cgi?id=270663</a>
<a href="https://rdar.apple.com/problem/124240255">rdar://problem/124240255</a>

Reviewed by Sammy Gill.

Indefinite items from other intrinsic tracks were not being taken
into account during the calculation.

masonry-align-content-003-expected-ref/expected.html:
Updated to exact pixel lengths, since &apos;auto&apos; can have different lengths in grid.

masonry-grid-template-columns-computed-withcontent.html:
Updated to have all intrinsic tracks being the same width.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/align-content/masonry-align-content-003-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/align-content/masonry-align-content-003-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent.html:
* Source/WebCore/rendering/Grid.cpp:
(WebCore::GridIterator::createForSubgrid):
* Source/WebCore/rendering/Grid.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::IndefiniteSizeStrategy::accumulateFlexFraction const):
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/277403@main">https://commits.webkit.org/277403@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66bb56cf3a83d07d42578287c7e56f98116e335f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47486 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50169 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43534 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49793 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32341 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24128 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38654 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48067 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40919 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19975 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21732 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42088 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5529 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52047 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22520 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45958 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23792 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41070 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44997 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/attributes (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10481 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24582 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23510 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/Grid.h:
```diff
@@ -89,6 +89,9 @@ class Grid final {
     unsigned maxRows() const { return m_maxRows; }
     unsigned maxColumns() const { return m_maxColumns; }
+
+    Vector<RenderBox*> autoMajorAxisAutoGridItems;
```
This should be a private variable that should only be accessible with whatever API we want to have for it. Also it should not be a Vector of raw pointers and should instead be a data structure that uses smart pointers.

I feel like ideally this should somehow be a part of the masonry layout class so that it can't get misued for other purposes, but if we don't want to do that I think it should be renamed to indicate clearly that this is being used as part of masonry layout (with the `m_` prefix to match to WebKit style guide). If we end up not moving it to `GridMasonryLayout` I wonder if it would make more sense for this state to live here or in the `GridTrackSizingAlgorithm` (I'm not sure)

stwrt replied:
Completely removing this code.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,9 +1355,41 @@ static LayoutUnit computeSubgridMarginBorderPadding(const RenderGrid* outermost,
 {
-    while (auto* gridItem = iterator.nextGridItem()) {
+    // This is for masonry items and helps track if we visited the item or not when going through this specific track.
+    HashMap<RenderBox*, bool> visitedGridItems;
```
Not sure I think this is the best name for this variable especially since we have to have a comment above that attempts to describe what it's being used for. We could probably get rid of the comment if we can somehow incorporate the informating it's trying to convey into the name. Also I think this could be a `WeakHashSet`

stwrt replied:
Updating to `visitedIndefiniteMasonryItemsInTrack` to give a better name.

smfr commented on Source/WebCore/rendering/Grid.h:
```diff
@@ -89,6 +89,9 @@ class Grid final {
     unsigned maxRows() const { return m_maxRows; }
     unsigned maxColumns() const { return m_maxColumns; }
+
+    Vector<RenderBox*> autoMajorAxisAutoGridItems;
```
Why this public member variable without an `m_` prefix? Also we don't use raw pointers any more.

stwrt replied:
Going to remove this and re-perform the computation in `resolveIntrinsicTrackSizes`.

smfr commented on Source/WebCore/rendering/Grid.h:
```diff
@@ -123,6 +126,8 @@ class GridIterator {
     static GridIterator createForSubgrid(const RenderGrid& subgrid, const GridIterator& outer);
 
     RenderBox* nextGridItem();
+    Vector<RenderBox*> masonryItems() { return m_grid.autoMajorAxisAutoGridItems; };
```
Does this need to copy the Vector? It should be `const`

stwrt replied:
Going to remove this and re-perform the computation in `resolveIntrinsicTrackSizes`.

smfr commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,9 +1355,41 @@ static LayoutUnit computeSubgridMarginBorderPadding(const RenderGrid* outermost,
 {
-    while (auto* gridItem = iterator.nextGridItem()) {
+    // This is for masonry items and helps track if we visited the item or not when going through this specific track.
+    HashMap<RenderBox*, bool> visitedGridItems;
```
Why not just a (Weak)HashSet?

stwrt replied:
Good point; updating to `SingleThreadWeakHashSet`.

smfr commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -1016,6 +1016,11 @@ void RenderGrid::placeItemsOnGrid(std::optional<LayoutUnit> availableLogicalWidt
 
+    if (isMasonry()) {
+        m_grid.m_currentGrid.get().autoMajorAxisAutoGridItems.clear();
+        m_grid.m_currentGrid.get().autoMajorAxisAutoGridItems.appendVector(autoMajorAxisAutoGridItems);
```
Why not just a setter that takes a Vector&&? But also you should try really hard to avoid having to store this vector of renderers. That always ends in tears.

stwrt replied:
Completely removed this code in next iteration.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1366,75 +1398,67 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+    for (auto* gridItem : iterator.masonryItems())
+        accumulateIntrinsicSizes(gridItem);
+
+    visitedGridItems.clear();
```
I don't think we need to clear this at the end of the function since it is local anyway

stwrt replied:
Removed.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1366,75 +1398,67 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        accumulateIntrinsicSizes(gridItem);
+
+    for (auto* gridItem : iterator.masonryItems())
+        accumulateIntrinsicSizes(gridItem);
```
We probably want to use `CheckedPtr` here

stwrt replied:
Updated to use CheckedPtr.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1353,87 +1353,118 @@ static LayoutUnit computeSubgridMarginBorderPadding(const RenderGrid* outermost,
 }
 
-void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track, unsigned trackIndex, GridIterator& iterator, Vector<GridItemWithSpan>& itemsSortedByIncreasingSpan, Vector<GridItemWithSpan>& itemsCrossingFlexibleTracks, SingleThreadWeakHashSet<RenderBox>& itemsSet, LayoutUnit currentAccumulatedMbp)
+void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track, unsigned trackIndex, GridIterator& iterator, Vector<GridItemWithSpan>& itemsSortedByIncreasingSpan, Vector<GridItemWithSpan>& itemsCrossingFlexibleTracks, SingleThreadWeakHashSet<RenderBox>& itemsSet, SingleThreadWeakListHashSet<RenderBox>& indefiniteItems, LayoutUnit currentAccumulatedMbp)
```
We really need to clean up this and `resolveIntrinsicTracks` at some point :/

sammygill replied:
Let's make it clear that this new argument is supposed to be used for masonry layout by indicating it in the name. We could also add a debug assert to make sure that this only contains item when the grid is in masonry as a sanity check

stwrt replied:
> We really need to clean up this and `resolveIntrinsicTracks` at some point :/

Wont' disagree with you there 🙃. I'll have a follow up patch to tackle that.

stwrt replied:
> Let's make it clear that this new argument is supposed to be used for masonry layout by indicating it in the name. We could also add a debug assert to make sure that this only contains item when the grid is in masonry as a sanity check

I'll update the name. I can add a check when entering `accumulateIntrinsicSizesForTrack` that the `indefiniteItems` should be empty.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1353,87 +1353,118 @@ static LayoutUnit computeSubgridMarginBorderPadding(const RenderGrid* outermost,
+    while (CheckedPtr<RenderBox> gridItem = iterator.nextGridItem())
+        accumulateIntrinsicSizes(gridItem.get());
+
+    for (auto it = indefiniteItems.begin(), end = indefiniteItems.end(); it != end; ++it) {
```
This also needs to be protected by a `CheckedPtr`

stwrt replied:
Updated to using a CheckedPtr.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1353,87 +1353,118 @@ static LayoutUnit computeSubgridMarginBorderPadding(const RenderGrid* outermost,
+void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track, unsigned trackIndex, GridIterator& iterator, Vector<GridItemWithSpan>& itemsSortedByIncreasingSpan, Vector<GridItemWithSpan>& itemsCrossingFlexibleTracks, SingleThreadWeakHashSet<RenderBox>& itemsSet, SingleThreadWeakListHashSet<RenderBox>& indefiniteItems, LayoutUnit currentAccumulatedMbp)
 {
-    while (auto* gridItem = iterator.nextGridItem()) {
+    SingleThreadWeakHashSet<RenderBox> visitedIndefiniteMasonryItemsInTrack;
```
I'm a little confused as to what this set is doing. I *think* I understand the intention based off of some of the comments below, but I"m not 100% sure (seems like to help force us to consider indefinite items over all of the intrinsically sized tracks). Is there a reason why we can't just consider if the item is contained within `indefiniteItems` since that should be the set of those items? I think this would be a good opportunity to provide a high level explanation of the overall architecture in the commit message to help provide a picture of what this patch is attempting to do.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1456,13 +1497,18 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
     Vector<GridItemWithSpan> itemsSortedByIncreasingSpan;
     Vector<GridItemWithSpan> itemsCrossingFlexibleTracks;
     SingleThreadWeakHashSet<RenderBox> itemsSet;
+    SingleThreadWeakListHashSet<RenderBox> indefiniteItems;
```
Same suggestion for the naming here since we'll be using it for a masonry specific purpose

stwrt replied:
Updated to masonryIndefiniteItems.

sammygill reviewed (approved):
Please make sure to export these test changes to WPT

bfgeek commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1353,87 +1355,118 @@ static LayoutUnit computeSubgridMarginBorderPadding(const RenderGrid* outermost,
+            isNewEntry = true;
+
+            // Correct the span as the grid item is coming from another track.
+            span = GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + span.integerSpan());
```
Hey Brandon - if an item is spanning multiple tracks, where do you account for it when it doesn't start in this track (say placed in the track before, but is spanning into this track)?

(I think this line is for when an item is starting in this track, but can't find the situation when it doesn't start in this track?)
(A few things can differ in track sizing due to this).

Ian

stwrt committed:
Masonry isn't performing intrinsic sizing correctly.
https://bugs.webkit.org/show_bug.cgi?id=270663
rdar://problem/124240255

Reviewed by Sammy Gill.

Indefinite items from other intrinsic tracks were not being taken
into account during the calculation.

masonry-align-content-003-expected-ref/expected.html:
Updated to exact pixel lengths, since 'auto' can have different lengths in grid.

masonry-grid-template-columns-computed-withcontent.html:
Updated to have all intrinsic tracks being the same width.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/align-content/masonry-align-content-003-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/align-content/masonry-align-content-003-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/masonry-grid-template-columns-computed-withcontent.html:
* Source/WebCore/rendering/Grid.cpp:
(WebCore::GridIterator::createForSubgrid):
* Source/WebCore/rendering/Grid.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::IndefiniteSizeStrategy::accumulateFlexFraction const):
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/277403@main
